hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCleaner.java

```diff
@@ -158,8 +159,7 @@ public HoodieCleaner(HoodieTable<T> hoodieTable, HoodieWriteConfig config) {
 
         // determine if we have enough commits, to start cleaning.
         if (commitTimeline.countInstants() > commitsRetained) {
-            HoodieInstant earliestCommitToRetain =
-                commitTimeline.nthInstant(commitTimeline.countInstants() - commitsRetained).get();
+            HoodieInstant earliestCommitToRetain = getEarliestCommitToRetain().get();
             List<List<HoodieDataFile>> fileVersions =
                 fileSystemView.getEveryVersionInPartition(partitionPath)
```

Comment:
Last time we dug into cleaner perf issue, it was fs.globStatus taking too much time or having a skew in partition distribution (ergo the Collection.shuffle above) . yeah might be good to lay out how much is the performance impact and how big the skew is. 

From what me & @kaushikd49 discussed offline, this was an urgent fix that was needed. Otherwise, its best to redo all this based on commit metadata, like how we redid the hive-sync now.